### DISCUSSION TITLE
Implementation of Qfile (quasar) format for local management

### DISCUSSION THREAD

**Discussion by vickorian** [zauberzeug/nicegui#269] [Ideas / Feature Requests]
For local implementation it feels that the Qfile format would allow for better and faster utilization without the need to wrangle the upload class. I work with data that would be nice to have a frontend on. Having the ability to select entire folders and then use python to retrieve the contents, and display them on the page would be great. Not to mention, i work with GBs of data and managing them with the uploader is beyond tedious when i really just need the filenames. I'm sure there are lots of people in the same boat.

**Comment by falkoschindler** [comment]
That sounds totally reasonable. But I'm not sure how QFile would help in contrast to the current `ui.upload` based on QUploader. The problem with any file selection element seems to be that it happens with JavaScript in the browser. And for security reasons the browser won't tell us the complete file path, only the name. Therefore we need to transfer the file content rather its path.

To be sure I quickly wrapped a QFile and subscribed to model changes. All I get is the file content.

```py
from nicegui import ui
from nicegui.element import Element

Element('q-file').on('update:model-value', lambda e: print(e, flush=True))

ui.run()
```

Related: https://github.com/zauberzeug/nicegui/issues/220, https://github.com/zauberzeug/nicegui/discussions/74

**Reply by vickorian** [reply]
The hope was to set up a queue system and use the file as a placeholder until it needed to be loaded and processed, then allow for download of the files as they came ready in batches. Running the uploader locally, i'm unable to access the files at my leisure and instead have to depend on the "on_upload" callback to perform an action. The hope was that with the url creation from the file api, it would let me make a proxy that could be accessed on demand. 

Currently the uploader isn't responding to UploadEventArguments so it makes naming impossible to do.

**Reply by falkoschindler** [reply]
I think it's a conceptual problem. As far as I understand, we can't get a filepath from the frontend.

But maybe you only need to fill a `ui.select` with the files of a certain well-known folder and the user can select. For multi-selection `ui.table` might also be useful.

What do you mean with "the uploader isn't responding to UploadEventArguments"?

**Reply by vickorian** [reply]
AttributeError: 'UploadEventArguments' object has no attribute 'files'

Get this when using the example for the AI interface from the examples list. Didn't know if it was a bug or not since i'm on Windows and not using the docker image.

I talked with one of my friends about the implementation and they said the only real way to handle it would be to make an API that interfaces with the gui. But that's an extraordinary amount of management for local files. I like the gui you've made though, but i don't think it fits my  use case unfortunately.

**Reply by falkoschindler** [reply]
> AttributeError: 'UploadEventArguments' object has no attribute 'files'

Sorry, there was an error in the event handler. After changing `ui.upload` and its event arguments recently, we overlooked to update this example.

**Reply by ustas4** [reply]
@falkoschindler  I'm sorry, but in latest version nice gui 3.2.0 exists issue UploadEventArguments has no attribute content, name . Tested in windows and Linux, chrome.

**Reply by falkoschindler** [reply]
In NiceGUI 3.0 the `UploadEventArguments` have changed. See the [migration guide](https://github.com/zauberzeug/nicegui/releases/tag/v3.0.0), point 12.

**Reply by ustas4** [reply]
Unfortunately, I tested the read and text functions to load a small txt file. Both functions e.file.read() and e.file.text(encoding='utf-8') return the error: nicegui: 'coroutine' object is not iterable

**Reply by evnchn** [reply]
@ustas4 Already replied in https://github.com/zauberzeug/nicegui/pull/5109#issuecomment-3478929607

Please do not comment twice. Thanks. I was second-guessing my memory so hard just now (did I lose my memory of my previous reply) 😵

**Comment by rodja** [comment]
#283 seems to be related.

**Comment by falkoschindler** [comment]
We will follow up on the idea of a local file selector over at #283.
